annotations/builder/src/main/java/io/sundr/builder/internal/functions/TypeAs.java

```diff
@@ -216,20 +219,32 @@ public ClassRef apply(TypeRef item) {
 
   public static final Function<TypeRef, TypeRef> UNWRAP_COLLECTION_OF = type -> {
     if (type instanceof ClassRef) {
-      ClassRef classRef = (ClassRef) type;
-      if (Collections.IS_COLLECTION.apply(classRef)) {
-        return classRef.getArguments().get(0);
-      }
+      return extractArgument((ClassRef) type, Collections.IS_COLLECTION, 0).orElse(type);
     }
     return type;
   };
 
+  private static Optional<TypeRef> extractArgument(ClassRef classRef, Function<TypeRef, Boolean> typeCheckFn,
+      int argumentIndex) {
+    if (typeCheckFn.apply(classRef)) {
+      if (classRef.getArguments().size() > argumentIndex) {
+        return Optional.of(classRef.getArguments().get(argumentIndex));
+      } else {
+        // Fallback
+        TypeDef typeDef = GetDefinition.of(classRef);
+        return Stream
```

Comment:
Since we go down that road, then maybe we should check the whole hierarchy instead of classes directly extended / implemented. 

It's possible that getting the `RichTypeDef` can easily address that.

Reply:
Thanks a lot for the feedback @iocanel , I have attempted to follow the suggestion but I might be missing some context (sorry I'm not really familiar with the codebase).

Let me try to understand:
 - what invariants are we missing by recursively iterating on extended/implemented classes? My understanding is that we are covering the full hierarchy
 - `RichTypeDef` doesn't seems to directly offer methods/helpers to traverse parents, do you have a reference or example of what you are thinking about?
 - The closest thing I found in the codebase is `Types.visitParents` but I couldn't get it to work as expected